sitemaps_from_robotstxt.py

```diff
@@ -0,0 +1,146 @@
+    sitemap_url_invalid_encoding = None
+    robots_txt_invalid_url = None
+    robots_txt_announcing_sitemap = None
+    robots_txt_with_more_than_50_sitemaps = None
+
+    def init_accumulators(self, session):
+        super(SitemapExtractorJob, self).init_accumulators(session)
+
+        sc = session.sparkContext
+        self.robots_txt_processed = sc.accumulator(0)
+        self.sitemap_urls_found = sc.accumulator(0)
+        self.sitemap_url_invalid = sc.accumulator(0)
+        self.sitemap_url_invalid_encoding = sc.accumulator(0)
+        self.robots_txt_invalid_url = sc.accumulator(0)
+        self.robots_txt_announcing_sitemap = sc.accumulator(0)
+        self.robots_txt_with_more_than_50_sitemaps = sc.accumulator(0)
+
+    @staticmethod
+    def reduce_group_by_key_func(kv: tuple):
+        """Map sitemap URL to cross-submit hosts:
+            sitemap_url => [host_1, ..., host_n]"""
+        sitemap_uri, hosts_lists = kv
+        try:
+            sitemap_host = urlparse(sitemap_uri).netloc.lower().lstrip('.')
+        except Exception as e:
+            raise RuntimeError("A invalid sitemap URI somehow made it through the initial parsing phase, this shouldn't happen: " + repr(e))
+
+        cross_submit_hosts = set()
+
+        for robots_txt_hosts in hosts_lists:
+            for robots_txt_host in robots_txt_hosts:
+                if robots_txt_host != sitemap_host:
+                    cross_submit_hosts.add(robots_txt_host)
+
+        return sitemap_uri, json.dumps(list(cross_submit_hosts))
+
+
+    def add_arguments(self, parser):
+        super(SitemapExtractorJob, self).add_arguments(parser)
+        # set output options to match old cc-mrjob output
+        parser.set_defaults(output_option=['sep=\t', 'escapeQuotes=false', 'header=false'])
+        parser.set_defaults(output_format='csv')
+
+
+    def process_record(self, record: ArcWarcRecord):
+        """ emit: sitemap_url => [host] """
+        if not self.is_response_record(record):
+            # we're only interested in the HTTP responses
+            return
+
+        self.robots_txt_processed.add(1)
+        url = None
+        host = None
+        n_sitemaps = 0
+
+        data = self.get_payload_stream(record).read()
+        for raw_line in data.splitlines():
+            raw_line = raw_line.strip()
+
+            match = SitemapExtractorJob.sitemap_pattern.match(raw_line)
+            if match:
+                sitemap_url = match.group(1).strip()
+                self.sitemap_urls_found.add(1)
+                n_sitemaps += 1
+                try:
+                    sitemap_url = sitemap_url.decode("utf-8", "strict")
+                except UnicodeDecodeError as e:
+                    self.get_logger().warn(f'Invalid encoding of sitemap URL {sitemap_url}: {repr(e)}')
+                    self.sitemap_url_invalid_encoding.add(1)
+                    continue
+
+                if not self._is_valid_url(sitemap_url, label_for_log='sitemap'):
```

Comment:
1. Does validation make sense before resolving the URL given the base URL (WARC-Target-URI)?
2. URL validation should be optional because
   - It's extra work (which is done anyway if you fetch the sitemaps)
   - We need to ensure that the [URL validator](https://github.com/python-validators/validators/blob/master/src/validators/url.py) does the same as the URL normalization and validation implemented in the crawler. This is a huge task, out-of-scope.